Fix #9509: Set empty user IDs for question suggestions to None instead of empty strings

sagangwee opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #9509.
2. This PR does the following:
- Sets empty user IDs for question suggestions to None instead of empty strings to fix issues with the user ID migration.
- Adds domain validations to suggestion objects checking for valid user ID formats.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

sagangwee committed:
Added domain validation checks for author and final_reviewer user IDs.

sagangwee committed:
Fix linting checks

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -162,6 +163,21 @@ def validate(self):
                     type(self.final_reviewer_id))
 
+        if (
+                not user_id_migration.verify_user_id_correct(self.author_id) and
```
@vojtechjelinek is this ok? (I thought we were going to delete user_id_migration after the migration.)

sagangwee commented:
@DubeySandeep @vojtechjelinek @nithusha21 User ID fixes for question suggestions. PTAL, thanks!

seanlip reviewed (commented):
Thanks for fixing this @sagangwee! LGTM from my point of view, just one question (might need @vojtechjelinek's input).

DubeySandeep commented on core/templates/pages/community-dashboard-page/services/question-suggestion.service.ts:
```diff
@@ -28,8 +28,8 @@ angular.module('oppia').factory('QuestionSuggestionService', [
-          assigned_reviewer_id: '',
-          final_reviewer_id: '',
+          assigned_reviewer_id: null,
+          final_reviewer_id: null,
```
I would suggest removing these two fields from the request for two reasons:
1. Backend should set it manually instead of depending on the client.
2. We don't expect the client to know their or someone else user_id.

DubeySandeep replied:
Ditto for the translation suggestion request. [Make sure to test exploration edit content suggestion flow if we are going to follow this pattern.]

sagangwee replied:
Done.

DubeySandeep commented on core/domain/suggestion_registry.py:
```diff
@@ -162,6 +163,21 @@ def validate(self):
                     type(self.final_reviewer_id))
 
+        if (
+                not user_id_migration.verify_user_id_correct(self.author_id) and
```
I don't think we should add this validation in the domain layer, I would rather suggest doing this in the service layer before storing the suggestion and in the controller layer while fetching request data. [Also, make sure to have this as a prod data validation.]

sagangwee replied:
I added a domain validation check in the services layer before the PUT. The domain validation will also be included in the prod data validation checks.

DubeySandeep commented on core/domain/suggestion_registry.py:
```diff
@@ -162,6 +163,21 @@ def validate(self):
                     'Expected final_reviewer_id to be a string, received %s' %
                     type(self.final_reviewer_id))
 
+        if (
```
Add `is not None` on line  161

sagangwee committed:
Removed user IDs from client-side suggestion creation calls

sagangwee committed:
Added is not None to check

vojtechjelinek commented on core/controllers/suggestion.py:
```diff
@@ -109,7 +109,7 @@ def post(self):
             self.user_id, self.payload.get('change'),
             self.payload.get('description'),
-            self.payload.get('final_reviewer_id'))
+            None)
```
Instead of this modify the `create_suggestion` to not receive `final_reviewer_id` and set it to `None` in the `create_suggestion`.

sagangwee replied:
Done.

vojtechjelinek commented on core/domain/suggestion_registry.py:
```diff
@@ -157,11 +158,26 @@ def validate(self):
+                    self.final_reviewer_id)
+        ):
+            raise utils.ValidationError(
+                'Expected final_reviewer_id to be in a valid user ID format.')
```
Maybe integrate this with the code on line 161 and have:
```python
if self.final_reviewer_id is not None:
    if not isinstance(self.final_reviewer_id, python_utils.BASESTRING):
        raise ...
    if not user_id_migration.verify_user_id_correct(self.final_reviewer_id):
        raise ...
```

sagangwee replied:
Done.

nithusha21 reviewed (approved):
I agree with Sandeep's review.. Nothing else to add!

vojtechjelinek reviewed (changes requested):
Thanks! Two comments.

sagangwee committed:
Fix cron_test

kevintab95 reviewed (approved):
LGTM for code-owner files, thanks @sagangwee!

sagangwee committed:
Add domain validation check before suggestion creation

sagangwee committed:
Indentation issues

sagangwee reviewed (commented):
@seanlip @vojtechjelinek @DubeySandeep PTAL. I removed the ability to specify final_reviewer_id from the client and added a domain validation check before the suggestion model PUT. The domain validation checks will also be included in the prod data validations. Thanks!

vojtechjelinek reviewed (approved):
LGTM!

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -156,11 +157,27 @@ def validate(self):
-            if self.final_reviewer_id:
+        if (
+                self.author_id is not None and
+                not user_id_migration.verify_user_id_correct(self.author_id)
```
@vojtechjelinek can you confirm that this is OK please? Will user_id_migration still be around after the migration / should we move this validation function to a more permanent place?

(I think you missed my comment [here](https://github.com/oppia/oppia/pull/9511#discussion_r439705012).)

seanlip replied:
(Ditto below)

vojtechjelinek replied:
It is okay since this is going to the release and I don't want to do any modifications to the migration code. We can move it away later. The migration will stay on the prod at least for month and when I will be removing the code I will move this verification elsewhere.

seanlip replied:
SGTM. Thanks!

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -245,7 +262,7 @@ class SuggestionEditStateContent(BaseSuggestion):
             self, suggestion_id, target_id, target_version_at_submission,
             status, author_id, final_reviewer_id,
-            change, score_category, last_updated):
+            change, score_category, last_updated=None):
```
Why the change to use a default arg (and why is it None)?

(Also, these values are supposed to be set at the datastore layer so I'm not sure why they're being passed in here.)

sagangwee replied:
One of the changes made in this PR was to create a domain object version of the suggestion before the initial datastore PUT so that we can call the `validate()` function on the domain object as a kind of additional data validation. Since before the initial PUT, there is no `last_updated`, I updated to the field to None here along the lines of usages I saw with other domain objects, e.g. with [Topic](https://github.com/oppia/oppia/blob/develop/core/domain/topic_domain.py#L449).

As to your second point, I agree, so if one wishes to create a domain version of a datastore object before an initial PUT, I thought it would make sense to allow the configuring of a None `last_updated`, but maybe one of the conventions with domain objects is that they should only represent existing datastore objects?

seanlip replied:
I think domain objects can sometimes represent objects that are yet to be put into the datastore, e.g. newly-created objects.

And you're right -- I checked existing domain objects and this does follow the existing convention, so I withdraw my comment. Thanks for explaining!

seanlip commented:
Hi @vojtechjelinek @aks681 PTAL, thanks! (@vojtechjelinek please note in particular the comments left for you)

kevintab95 commented:
Hi @DubeySandeep, @sagangwee, do we need to enable the community dashboard with this fix?

seanlip commented:
Not yet. All bugs in our doc need to be addressed prior to the contributor dashboard being re-enabled.

seanlip commented:
(Also note that the current name is wrong and should be changed -- I think this is mentioned in the doc outlining the current issues. It should be "contributor" dashboard, not community dashboard.)

DubeySandeep commented:
> Not yet. All bugs in our doc need to be addressed prior to the contributor dashboard being re-enabled.


Also, we have planned to enable it after adding a feature to allow contributors to add images in the suggestion!

kevintab95 commented:
@sagangwee please mind the backend test failures!

sagangwee committed:
Provide valid exploration category in save_new_default_exploration test_utils method

sagangwee committed:
Fix A category tests

sagangwee reviewed (commented):
@DubeySandeep @seanlip Addressed reviewer comments. PTAL. Thanks!

seanlip reviewed (approved):
LGTM, can't see any other issues. Thanks @sagangwee!

seanlip commented:
Out of curiosity, what was the issue with "A category"? Did it fail validation? (If so, curious why it passed before since it looks like a generic utility method.)

seanlip commented:
@aks681 PTAL as codeowner. Thanks!

sagangwee commented:
@seanlip Yeah with my addition of calling the suggestion domain object's `validate` in `create_suggestion`, some existing tests would fail a [validation check](https://github.com/oppia/oppia/blob/develop/core/domain/suggestion_registry.py#L188) of the score category field. I just chose one of the valid categories arbitrarily for the affected tests.

seanlip commented:
Ah, ok. Sg, thanks!

aks681 reviewed (approved):
Lgtm as codeowner.

sagangwee committed:
Fix reviewable suggestion tests

DubeySandeep commented on core/controllers/suggestion.py:
```diff
@@ -108,8 +108,7 @@ def post(self):
             self.payload.get('target_type'), self.payload.get('target_id'),
```
Can you please add a check for theses values above?

sagangwee replied:
The approach I took was to marshal the values into the domain object and call `validate` before the datastore PUT, which will occur during the `create_suggestion` call.

DubeySandeep replied:
But that if validation fails we will throw a 500 error but we should throw a 400 error for invalid input right?

sagangwee replied:
I see. I updated this to catch ValidationErrors and throw InvalidInput 400 errors eternally to the user. Does that work?

DubeySandeep commented on core/controllers/suggestion_test.py:
```diff
@@ -1090,9 +1087,6 @@ def setUp(self):
         # Login and create exploration and suggestions.
         self.login(self.EDITOR_EMAIL)
-        self.save_new_linear_exp_with_state_names_and_interactions(
-            self.EXP_ID, self.editor_id, ['State 1', 'State 2', 'State 3'],
```
Why this changes?

sagangwee replied:
While fixing issues with this test during development, I realized that this line doesn't actually do anything since the exploration is not published (an exploration with the same ID is published above this line).

DubeySandeep commented on core/domain/suggestion_registry.py:
```diff
@@ -156,11 +157,27 @@ def validate(self):
-            if self.final_reviewer_id:
+        if (
+                self.author_id is not None and
+                not user_id_migration.verify_user_id_correct(self.author_id)
```
Don't add this service layer checks here, should we move this to suggestion_service before creating and saving suggestions?

sagangwee replied:
These checks will be made in suggestion_service before creating suggestions, as the domain object is created and its `validate` method is called before modifying the datastore.

DubeySandeep reviewed (commented):
LGTM for the code owner files, just a few minor comments/questions, PTAL!

DubeySandeep reviewed (changes requested):
(Marking this request changes)

sagangwee commented:
@DubeySandeep Addressed your comments. PTAL. The approach I opted for was to create a domain version of the suggestion in suggestion_services during creation and call its `validate` method there. I think that accomplishes your request to validate at the service level, but let me know what you think. Thanks!

sagangwee committed:
Catch ValidationErrors as InvalidInput

sagangwee committed:
Remove unused variable

sagangwee commented:
@DubeySandeep Made some changes according to your comments. PTAL. Thanks!

DubeySandeep reviewed (approved):
LGTM!

sagangwee committed:
Resolve merge conflicts

sagangwee committed:
Fix argument length lint check

sagangwee committed:
Merge remote-tracking branch 'upstream/develop' into final-reviewer-id-null-fix

sagangwee committed:
fix too many positional arguments

sagangwee committed:
Merge remote-tracking branch 'upstream/develop' into final-reviewer-id-null-fix

iamprayush commented:
@sagangwee The backend tests are failing, PTAL, thanks!

sagangwee committed:
Merge remote-tracking branch 'upstream/develop' into final-reviewer-id-null-fix

sagangwee committed:
Removed extra argument to create_suggestion
